### PR TITLE
Fix hessian bug

### LINKED ISSUE #1231: Hessian failure for logjoint
MWE:

```julia
​
using Turing
using DynamicPPL
using ForwardDiff
​
@model function tst(x, ::Type{TV}=Vector{Float64}) where {TV}
	params = TV(undef, 5)
	params .~ Normal(0, 1)
​
	x ~ MvNormal(params, 1)
end
​
function make_logjoint(model::DynamicPPL.Model, ctx::DynamicPPL.AbstractContext)
    # setup
    varinfo_init = Turing.VarInfo(model)
    spl = DynamicPPL.SampleFromPrior()    
    DynamicPPL.link!(varinfo_init, spl)
​
    function logπ(z; unlinked = false)
        varinfo = DynamicPPL.VarInfo(varinfo_init, spl, z)
​
        unlinked && DynamicPPL.invlink!(varinfo_init, spl)
        model(varinfo, spl, ctx)
        unlinked && DynamicPPL.link!(varinfo_init, spl)
​
        return -DynamicPPL.getlogp(varinfo)
    end
​
    return logπ
end
​
data = randn(5)
model = tst(data)
​
likelihood = make_logjoint(model, DynamicPPL.LikelihoodContext())
target(x) = likelihood(x, unlinked=true)
​
H = ForwardDiff.hessian(target, zeros(5))
```

Error: 

```julia
ERROR: LoadError: TypeError: in typeassert, expected Float64, got ForwardDiff.Dual{Nothing,ForwardDiff.Dual{Nothing,Float64,5},5
}
Stacktrace:
 [1] setindex!(::Array{Float64,1}, ::ForwardDiff.Dual{ForwardDiff.Tag{typeof(target),Float64},ForwardDiff.Dual{ForwardDiff.Tag{t
ypeof(target),Float64},Float64,5},5}, ::Int64) at ./array.jl:825
 [2] copyto! at ./multidimensional.jl:962 [inlined]
 [3] copyto! at ./broadcast.jl:905 [inlined]
 [4] copyto! at ./broadcast.jl:864 [inlined]
 [5] materialize! at ./broadcast.jl:823 [inlined]
 [6] dot_assume(::SampleFromPrior, ::NoDist{Univariate,Continuous,Normal{Float64}}, ::Array{VarName{:params,Tuple{Tuple{Int64}}}
,1}, ::Array{Float64,1}, ::VarInfo{NamedTuple{(:params,),Tuple{DynamicPPL.Metadata{Dict{VarName{:params,Tuple{Tuple{Int64}}},Int
64},Array{Normal{Float64},1},Array{VarName{:params,Tuple{Tuple{Int64}}},1},Array{ForwardDiff.Dual{ForwardDiff.Tag{typeof(target)
,Float64},ForwardDiff.Dual{ForwardDiff.Tag{typeof(target),Float64},Float64,5},5},1},Array{Set{DynamicPPL.Selector},1}}}},Forward
Diff.Dual{ForwardDiff.Tag{typeof(target),Float64},ForwardDiff.Dual{ForwardDiff.Tag{typeof(target),Float64},Float64,5},5}}) at /$
ome/cameron/.julia/dev/DynamicPPL/src/context_implementations.jl:259
 [7] _dot_tilde at /home/cameron/.julia/dev/DynamicPPL/src/context_implementations.jl:222 [inlined]
 [8] dot_tilde at /home/cameron/.julia/dev/DynamicPPL/src/context_implementations.jl:166 [inlined]
 [9] dot_tilde_assume at /home/cameron/.julia/dev/DynamicPPL/src/context_implementations.jl:200 [inlined]
 [10] macro expansion at /home/cameron/.julia/dev/Misc/hessian-broken.jl:8 [inlined]
 [11] ##evaluator#465(::Model{var"###evaluator#465",(:x, :TV),Tuple{Array{Float64,1},Type{Array{Float64,1}}},(),ModelGen{var"##$
generator#466",(:x, :TV),(:TV,),Tuple{Type{Array{Float64,1}}}}}, ::VarInfo{NamedTuple{(:params,),Tuple{DynamicPPL.Metadata{Dict$
VarName{:params,Tuple{Tuple{Int64}}},Int64},Array{Normal{Float64},1},Array{VarName{:params,Tuple{Tuple{Int64}}},1},Array{Forwar$
Diff.Dual{ForwardDiff.Tag{typeof(target),Float64},ForwardDiff.Dual{ForwardDiff.Tag{typeof(target),Float64},Float64,5},5},1},Arr$
y{Set{DynamicPPL.Selector},1}}}},ForwardDiff.Dual{ForwardDiff.Tag{typeof(target),Float64},ForwardDiff.Dual{ForwardDiff.Tag{type$
f(target),Float64},Float64,5},5}}, ::SampleFromPrior, ::LikelihoodContext{Nothing}) at /home/cameron/.julia/dev/DynamicPPL/src/$
ompiler.jl:381
 [12] Model at /home/cameron/.julia/dev/DynamicPPL/src/model.jl:139 [inlined]
```

### PR DESCRIPTION
**Pull request by cpfiffer**
Fixes #1231, which is necessary for #1230. Adds a test to make sure the Hessian is being correctly calculated with ReverseDiff and ForwardDiff.

### CODE REVIEW

**Comment by yebai** [comment]
@cpfiffer It seems some tests are broken with this PR

**Comment by cpfiffer** [comment]
Yeah, it requires https://github.com/TuringLang/DynamicPPL.jl/pull/83 to be merged & released before it should work, though I'm not sure what stuff is in the air over there. @mohamed82008 could https://github.com/TuringLang/DynamicPPL.jl/pull/83 be merged and released without too much issue at DynamicPPL?

**Comment by cpfiffer** [comment]
The test failures here seem unrelated.

**Comment by devmotion** [comment]
I think they're not actually. Using `.~` inside a model will lead to parsing errors on Julia 1.0. Alternatively, you could use `@.` and escape non-broadcasted parts with `$`, or use some function like syntax (`~.(LHS, RHS)` IIRC).

**Comment by devmotion** [comment]
See, e.g., https://github.com/TuringLang/DynamicPPL.jl/blob/c3d430c774628e1d7f5f8ee0167c0704a84df265/test/compiler.jl#L365 and https://github.com/TuringLang/DynamicPPL.jl/commit/43e8eeebaf41189e1f305e6880dd14c7ef18da86#diff-692d6d33d7ed80773e898f56e60b88ab (for the functional alternative `(.~)(LHS, RHS)`).

File: test/core/ad.jl [hunk 1]
```diff
@@ -302,4 +302,42 @@ _to_cov(B) = B * B' + Matrix(I, size(B)...)
         Turing.setadbackend(:zygote)
         sample(invwishart(), HMC(0.01, 1), 1000);
     end
+    @testset "Hessian test" begin
+        @model function tst(x, ::Type{TV}=Vector{Float64}) where {TV}
+            params = TV(undef, 2)
+            params .~ Normal(0, 1)
```
Comment (cpfiffer) [inline test/core/ad.jl:308]:
```suggestion
            @. params ~ Normal(0, 1)
```

**Comment by cpfiffer** [comment]
Oh, I see what you mean. I hadn't noticed that this syntax didn't work for Julia 1.0. Let me try this and see what happens.

**Comment by cpfiffer** [comment]
Alright, looks like tests are passing.

**Pull request merged**